source/docs/software/vision-processing/apriltag/apriltag-intro.rst

```diff
@@ -0,0 +1,237 @@
+
+      * Calculate the expected interior pixel coordinates where the center of each bit should be
+      * Mark each location as "1" or "0" by comparing the pixel intensity to a threshold
+      * Find the tag ID which most closely matches what was seen in the image, allowing for one or two bit errors.
+
+      It is possible there is no valid tag ID which matches the suspect tag. In this case, the decoding process stops.
+
+   .. tab:: Fit External Quad
+
+      .. image:: images/fit_ext_quad.png
+         :alt: External quadrilateral fitting process applied to each target
+
+      Now that we have a tag ID for the region of pixels, we need to do something useful with it.
+
+      For most FRC applications, we care about knowing the precise location of the corners of the tag, or its center. In both cases, we expect the resolution-lowering operation we did at the beginning to have distorted the image, and we want to undo those effects.
+
+      The algorithm to do this is:
+
+      * Use the detected tag location to define a region of interest in the original-resolution image
+      * Calculate the :term:`gradient` at pre-defined points in the region of interest to detect where the image most sharply transitions between black to white
+      * Use these gradient measurements to rapidly re-fit an exterior quadrilateral at full resolution
+      * Use geometry to calculate the exact center of the re-fit quadrilateral
+
+      Note that this step is optional, and can be skipped for faster image processing. However, skipping it can induce significant errors into your robot's behavior, depending on how you are using the tag outputs.
+
+
+Usage
+-----
+
+2D Alignment
+^^^^^^^^^^^^
+
+A simple strategy for using targets is to move the robot until the target is centered in the image. Assuming the field and robot are constructed such that the gamepiece, scoring location, vision target, and camera are all aligned, this method should proved a straightforward method to automatically align the robot to the scoring position.
+
+Using a camera, identify the _centroid_ of the apriltags in view. If the tag's ID is correct, apply drivetrain commands to rotate the robot left or right until the tag is centered in the camera image.
+
+This method does not require calibrating the camera or performing the homography step.
+
+
+3D Alignment
+^^^^^^^^^^^^
+
+.. image:: images/homography.png
+
+.. todo:: Code examples coming soon!
+
+A more advanced usage of AprilTags is to use their corner locations to help perform on-field localization.
+
+Each image is searched for AprilTags using the algorithm described on this page. Using assumptions about how the camera's lense distorts the 3d world onto the 2d array of pixels in the camera, an estimate of the camera's position relative to the tag is calculated.
```

Comment:
Maybe a line about calibration? In 2022, teams oculd use corner data of bounding boxes to get pose estimation for the 2022 goal, they cant really get pose estimation with just corner cordinates in 2d mode (photon probably doesnt report these but the connotation between corner points and getting information without calibration exists. If that sentence makes any sense